[MIG][15.0] sale_stock_sale_management: add migration for 15.0

daonamutc1 opened the pull request:
- [x] [sale_management](https://viindoo.com/web#id=23090&action=409&active_id=177&model=project.task&view_type=form&cids=1&menu_id=289)
- [x] [sale_stock](https://viindoo.com/web#id=23099&action=409&active_id=177&model=project.task&view_type=form&cids=1&menu_id=289)

- Convert field note from text to html
- Change effective_date from date to datetime
- Rename field product_packaging (sale_stock) to product_packaging_id (sale)

Video:
- Before: https://drive.google.com/file/d/1k7pWllMLVrOK5tnr8TIAbCY5ntaJFSDm/view?usp=sharing
- After: https://drive.google.com/file/d/1pL-ABlkIt9Cy6Puz5Gr_YCuRrKf13zhW/view?usp=sharing

daonamutc1 committed:
[MIG] sale_management: add migration for 15.0
- Convert field note from text to html

royle-vietnam commented on openupgrade_scripts/scripts/sale_stock/15.0.1.0/post-migration.py:
```diff
@@ -0,0 +1,18 @@
+# Copyright 2021 ForgeFlow S.L.  <https://www.forgeflow.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
```
@daonamutc1 xóa đoạn này đi anh nhé

royle-vietnam commented on openupgrade_scripts/scripts/sale_stock/15.0.1.0/post-migration.py:
```diff
@@ -0,0 +1,18 @@
+from openupgradelib import openupgrade
+
+
+def date_to_datetime_fields(env):
+    openupgrade.date_to_datetime_tz(
```
@daonamutc1  dùng hàm `date_to_datetime_tz` sẽ không đúng với logic của hàm `_compute_effective_date`
```python
def _convert_date_to_datetime_sale_order_effective_date(env):
    # convert column from date to datetime
    openupgrade.logged_query(
            env.cr,
            """
            ALTER TABLE sale_order ALTER COLUMN effective_date timestamp without time zone
           """,
    )
    # convert value from date to datetime
    openupgrade.logged_query(
            env.cr,
            """
            # Chỗ này cần lấy id của sale_order và date_done của stock_pikcing nhỏ nhất
            WITH subquery(
            )
           # Cập nhật effective_date theo giá trị từ subquery
          UPDATE sale_order so
          SET effective_date = sub.effective_date
          FROM subquery sub
          WHERE sub.so_id = so.id
           """,
    )
```

royle-vietnam commented on openupgrade_scripts/scripts/sale_stock/15.0.1.0/pre-migration.py:
```diff
@@ -0,0 +1,37 @@
+            env.cr,
+            """
+            ALTER TABLE sale_order_line
+            ADD COLUMN product_packaging_id integer
```
@daonamutc1 dùng `ADD COLUMN IF NOT EXISTS` sẽ đơn giản hơn

royle-vietnam commented on openupgrade_scripts/scripts/sale_stock/15.0.1.0/upgrade_analysis_work.txt:
```diff
@@ -0,0 +1,14 @@
+sale_stock   / sale.order               / effective_date (date)         : type is now 'datetime' ('date')
+# DONE: post-migration: set datetime values with the user tz into account
+
+sale_stock   / sale.order.line          / product_packaging (many2one)  : DEL relation: product.packaging
```
@daonamutc1 cái này chỉ cần sử dụng hàm `rename_columns` là xong anh nhé

royle-vietnam replied:
@daonamutc1 nhân tiện, anh sửa lại file `upgrade_analysis.txt` của `sale` từ `Nothing to do` thành `# NOTHING TO DO: handled in  sale_stock module (renamed column from product_packaging to product_packaging_id)`

daonamutc1 replied:
Vì odoo chuyển qua module sale. Module sale lại được update trước module này, nên cột product_packaging_id đã có trước khi update module sale_stock. Chỗ này buộc phải copy giá trị qua.

royle-vietnam commented on openupgrade_scripts/scripts/sale_stock/15.0.1.0/pre-migration.py:
```diff
@@ -0,0 +1,37 @@
+from openupgradelib import openupgrade
+
+
+def _copy_effective_date_columns(env):
+    openupgrade.copy_columns(
+        env.cr,
+        {
+            "sale_order": [
+                ("effective_date", None, None),
+            ]
+        },
+    )
+
```
@daonamutc1 copy trường này ra cũng không có ý nghĩa gì, nếu migration sai, có thể dựa vào date_done của stock.pikcing để sửa
```suggestion
```

daonamutc1 replied:
Anh test lại rồi đẩy code sau nhé

daonamutc1 committed:
[MIG] sale_stock: add migration for 15.0
- Change effective_date from date to datetime
- Rename field product_packaging (sale_stock) to product_packaging_id (sale)

royle-vietnam commented on openupgrade_scripts/scripts/sale_stock/15.0.1.0/upgrade_analysis_work.txt:
```diff
@@ -0,0 +1,14 @@
+---Models in module 'sale_stock'---
+---Fields in module 'sale_stock'---
+sale_stock   / sale.order               / effective_date (date)         : type is now 'datetime' ('date')
+# DONE: post-migration: set datetime values with the user tz into account
```
```suggestion
# DONE: post-migration: set effective_date = min(pickings.date_done)
```
```suggestion
# DONE: post-migration: set datetime values with the user tz into account
```

royle-vietnam replied:
```suggestion
# DONE: post-migration: set effective_date = min(pickings.date_done)
```

royle-vietnam committed:
Update openupgrade_scripts/scripts/sale_stock/15.0.1.0/upgrade_analysis_work.txt
